### PR TITLE
feat: add v2 endpoints controller

### PR DESCRIPTION
**Pull request by zalimeni**
_Follows pattern of iterative review in https://github.com/hashicorp/consul-k8s/pull/2868, and currently targets that branch due to dependence on common code added there_

------------

This PR is part of a series of PRs that add functionality to the V2 Endpoints Controller. For the MVP task to be complete, there will be a few additional changes to resolve TODOs, debounce upsert writes if needed, and add further tests.

Changes proposed in this PR:

- Add the V2 Endpoints Controller
- Add enough logic in the controller to manage Consul Services
- Endpoints controller test framework, borrowed from linked Pod Controller PR (thanks @DanStough!)

How I've tested this PR: Unit tests

How I expect reviewers to test this PR: 👁️‍🗨️👃🏻👁️‍🗨️ 

Checklist:
- [x] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: control-plane/connect-inject/common/common_test.go
Comment (zalimeni) [inline control-plane/connect-inject/common/common_test.go:323]: Full disclosure, I didn't test panics in this function intentionally because 1) I'm not really sure how to trigger one easily and 2) the purpose of the function is use when doing things that should never panic.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:151]: [This](https://docs.google.com/document/d/1q01OHEnQ6Ir40y5GeVq53X5iN-5YExHshX_pashLvlQ/edit?disco=AAAA2PmcWks) 🧵 discusses maybe not needing explicit exclusion for services. I'm not 100% certain whether this controller should expect any opt-outs that will impact service registration, outside of disabled k8s namespaces from config, already checked above.

Input welcome! Lacking a strong opinion, I'll probably convert this to a follow-up ticket as a placeholder for a pre-GA decision.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:184]: Feel free to correct me if anyone knows better here. 🧵 for context [here](https://hashicorp.slack.com/archives/C04MG2UT3PA/p1692203149392719?thread_ts=1692199919.770389&cid=C04MG2UT3PA).

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:213]: My thinking on this r/n is that we should just try to read the existing Service, and if we find it, check for total equality minus resource version. If there's a diff, rewrite.

If we move to an `EndpointSlices`/`Pods`-based approach in the future, we'll need to do some sort of checking for stale workload selector values, but since `Endpoints` gives us a full snapshot up to the 1K pod limit, we should be able to just upsert today w/o being concerned with drift.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:254]: I'd appreciate a spot check on this mapping JIC I'm missing some nuance - cc @ishustava in case you have input

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:264]: Assuming we support discovery sync (non-mesh use cases) in the future, I think we'd probably want to exclude this port addition - but it's probably something we can defer. Marked `TODO(zalimeni)` JIC we already know the answer before I ticket it as a follow-up.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:351]: Maybe overkill, maybe share-able 🤷🏻‍♂️

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go:356]: @DanStough you might be interested in adding this to your comp fn - it's what we use in goldens and is really nice for a visual diff

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go:367]: `TODO`: I'll need to fix this for non-default ns tests

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:296]: I wasn't sure whether we wanted to keep allowing for custom meta via annotations, but seemed feasible to adapt from the v1 controller. Thoughts? Happy to cut if we're not sure about supporting it, or the interpolation of `$SERVICE_NAME`.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:212]: Was there a specific case you were trying to address by taking the diff and rewriting it? There was some discussion of running multiple versions of the controller concurrently in the future, and I wonder if this could create flapping with different workers fighting over the same service.

I think I naively assumed that if the write doesn't error, the resource service has persisted the service. Some future work is planned for later releases to sync the status of Consul resources back to Kubernetes resources. Do you think this would address any of your concerns?

File: control-plane/connect-inject/common/common.go
Comment (DanStough) [inline control-plane/connect-inject/common/common.go:154]: I wonder if we technically need to include prefixes with some of these protocols. I was reading the API docs for [appProtocol](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#serviceport-v1-core) and it mentions only[ IANA names](https://www.iana.org/assignments/service-names) can be unprefixed. Only my first time seeing this so I could be mistaken.

I also think an annotation on the service or pod might be easier for customers to support. e.g. the [Bitnami Postgres helmchart](https://github.com/bitnami/charts/tree/main/bitnami/postgresql) exposes annotations for services and certain pods but not a change of appProtocol.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:295]: I left this out for now, but I don't have a strong opinion.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:253]: I don't think `targetPort` will work in the case it is a number, but the pod defines the service with a name. We will need to look up the corresponding port name in the pod.  I'm open to changing how this works on the Pod side but not sure how you get around the name being optional and `targetPort` working on either.

**Review by DanStough — COMMENTED** [review]
I didn't get a chance to review the test case yet, but everything else looks great! I wanted to send you my open-ended comments in advance, like how `targetPort` works and the `appProtocol`.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:212]: > Was there a specific case you were trying to address by taking the diff and rewriting it?

Mainly just avoiding needless writes - TBF I'm not fully up to speed on the resource API and state store changes, but my basic understanding is reads are cheaper, and we are theoretically capped in the 2-3K/s for writes. This seemed like an easy way to avoid needless clobbering, esp. when potentially many reconciles loops attempt to write at once w/o versions and the baked-in CAS retries compete w/ each other.

>  There was some discussion of running multiple versions of the controller concurrently in the future, and I wonder if this could create flapping with different workers fighting over the same service.

I think the flapping would occur regardless of pre-reads? But that's a good callout. I'd hope we'd come up w/ some one-way version tagging similar or identical to the `managed-by` tag if needed to avoid issues from that. (Do you have a link to that conversation by chance?)

> I think I naively assumed that if the write doesn't error, the resource service has persisted the service. Some future work is planned for later releases to sync the status of Consul resources back to Kubernetes resources. Do you think this would address any of your concerns?

That's my assumption too 👍🏻 ATM I'm only thinking of debouncing, though bidirectional sync would add complexity to that.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:253]: > I don't think targetPort will work in the case it is a number, but the pod defines the service with a name

(_This might be wrong or obvious, feel free to correct me_) It seems to me from the k8s [docs](https://kubernetes.io/docs/concepts/services-networking/service/#defining-a-service) that there's an inherent tradeoff that applies generally: you need to pick whether to use a name consistently across service and pods (which enables mixed pod port numbers under one name), or you need to pick a consistent pod port number and use that in the service. Either way, it seems selected pods need to adhere to the service's chosen int-or-string value for `targetPort`.

Our docs for Consul `Service.TargetPort` say `target_port is the name of the workload port`, so I took that to mean we defer totally to the value used by the k8s service.

It seems wrong to me to reach into the pods to infer anything beyond the workload selector, since the `targetPort` mapping isn't really a `consul-k8s`-specific limitation... But I might be missing a reason we'd want to do that.

> I'm open to changing how this works on the Pod side

What did you have in mind here?

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:212]: I think it was in a casual Zoom, so I don't happen to have a link to the convo.

File: control-plane/connect-inject/common/common_test.go
Comment (DanStough) [reply control-plane/connect-inject/common/common_test.go:323]: Let's keep it for now.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:253]: > It seems wrong to me to reach into the pods to infer anything beyond the workload selector, since the targetPort mapping isn't really a consul-k8s-specific limitation

Yeah, I don't like the idea of reaching in either. The pods may be slightly different, and the names don't align, making the lookup challenging. But I think I disagree that there isn't something K8s-specific here. We generate a Consul port name per port, even if one wasn't defined in the Kube spec, and Kubernetes doesn't necessarily require you to use any name at all, so there is a mismatch of strong (consul) vs. weak (k8s) references. I think I had the opposite reaction that since the VM case is straightforward, we'd have to "fix it" in consul-k8s.

>> I'm open to changing how this works on the Pod side
No specific ideas here. I'm just trying not to look like I'm suggesting a singular solution.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go]: Asked 🧵 [here](https://hashicorp.slack.com/archives/C04AM35BFQT/p1694005306847929) to seek wisdom

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:258]: Started 🧵 [here](https://hashicorp.slack.com/archives/C04MG2UT3PA/p1694005024255509) to figure out how we want this to work

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go]: Marking this as a figure-out-later item for now

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go]: Omitted for brevity 🚀 TODO'd for follow-up

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go:854]: Leaving as follow-up in the interest of early testing using default ns

**Review by zalimeni — COMMENTED** [review]
@DanStough @thisisnotashwin ok, I think this is ready for a next pass. I've added more tests around the essential CrUD bits as well as the pod fetching behavior since that seemed valuable to lock in.

Assuming no major changes are needed, I'll plan to merge this to unblock testing while I follow up on remaining TODOs and Slack 🧵s.

File: control-plane/connect-inject/common/common_test.go
Comment (zalimeni) [reply control-plane/connect-inject/common/common_test.go:323]: Added handling for `anypb.Any` input -> return unmodified and test case to match.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (DanStough) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:192]: Just an interesting fact, I noticed there is a field [generateName](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.26/#objectmeta-v1-meta) that includes the shared prefixes that pods used. I think the current approach is more robust though.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go
Comment (DanStough) [inline control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go:1]: WDYT about an additional test with some `NotReadyAddresses` in the Endpoint Subset to test that they are included in the name generation?

**Review by DanStough — APPROVED** [review]
I think based on all the tradeoffs we discussed to get a version of this ready to test, LGTM.

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go:192]: Ah yeah, thanks for pointing out! I think I ran across some doc referring to that field early on and had the same thought; it didn't seem guaranteed enough to hang implementation on over the owner ref.

For fun, here's an example:
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2023-08-28T15:55:38Z"
  generateName: fluentd-elasticsearch-85996f8dbd-
  labels:
    app: nginx
    pod-template-hash: 85996f8dbd
  name: fluentd-elasticsearch-85996f8dbd-9r67q
  namespace: default
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: fluentd-elasticsearch-85996f8dbd
    uid: 610a6961-944c-4863-9738-7d783dd1992c
  resourceVersion: "1841"
  uid: 28e1fdd5-0107-4ddc-890b-e07065d6000f
```

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller_test.go:1]: Great call, I'll do that 👍🏻

**Review by thisisnotashwin — APPROVED** [review]
BEAUTIFUL!!!

File: control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go
Comment (zalimeni) [reply control-plane/connect-inject/controllers/endpointsv2/endpoints_controller.go]: Summary from Slack: not entirely conclusive, but we agree early optimization should be generally avoided until we get a better idea of write performance in the wild.

I think it's fair to argue that debouncing rapid identical writes falls under low-hanging fruit to reduce needless load, but is leading me to additional questions about how we'll handle state in resources that may be mutated by multiple sources/data flows (not sure how much of this there will be - could resource `Meta` tagging be an example?), since we also can't just blindly rewrite records in that scenario.

Regardless, this is tagged for follow-up outside this PR.

File: control-plane/connect-inject/common/common.go
Comment (zalimeni) [reply control-plane/connect-inject/common/common.go:158]: Added this as a bullet in follow-up ticket so we don't lose track but probably don't need to figure out right this second.

Comment (zalimeni) [reply control-plane/connect-inject/common/common.go:158]: Circling back to this one to tie up the loose end: @ishustava, do you happen to have perspective on either of Dan's questions above?

My gut is to leave the unprefixed protocols in place for now since there seems to be precedent in other meshes, and wait to see if we need annotations in addition to just `appProtocol`. If that's not ideal, we can fix now or plan to follow up in 1.18.

**Pull request merged**
